### ISSUE TITLE
unable to get a exit code 0 from docker run command

### ISSUE THREAD

**Issue by ibrahimabedalghafer** [vmware/vic#7345]
Hi all,

I am new on the vmware integrated container, and I cannot find a support fourm to express my issue, so please if I am hitting the wrong place please accept my apologize and please guide to where from I can get some support

when I run the command 
docker run ubuntu 
I get the below message :
docker: Error response from daemon: Server error from portlayer: unable to wait for process launch status: Post https://<vcenter IP>/sdk: context deadline exceeded
and when open a console of the container I get the below message stuck on it 
INFO waiting clear signal to launch ........

I am having esxi 6.5 and vcenter 6.5 U1e and deployed over a physical machine

**Comment by ibrahimabedalghafer** [comment]
Dear Matthew,

please find below:
- I am using IBM x3650 M4 server with 48GB of RAM , and it is a single host in the cluster with DRS enabled
- I am having datastore [1-5] are local disks attached to the server, and two iscsi datastores
- the image store and the volume store are same datastore that is one of iscsi datastore
- I am deploying VIC 1.3.1, and the below command is used for deploy the VCH
vic-machine-linux create --name vic-m1 \
--target <Vcenter IP> \
--user "administrator@vsphere.local" \
--password <admin password> \
--compute-resource sts-lab-cluster \
--thumbprint <thumbprint> \
--image-store vic-datastore --volume-store vic-datastore:default \
--bridge-network bridge-dpg --public-network external-dpg --container-network external-dpg \
--no-tlsverify

and below is the ouput 
INFO[0000] ### Installing VCH ####                      
WARN[0000] Using administrative user for VCH operation - use --ops-user to improve security (see -x for advanced help) 
INFO[0000] Loaded server certificate vic-m1/server-cert.pem 
WARN[0000] Configuring without TLS verify - certificate-based authentication disabled 
INFO[0000] Validating supplied configuration            
INFO[0000] vDS configuration OK on "bridge-dpg"         
INFO[0000] vDS configuration OK on "external-dpg"       
INFO[0000] vCenter settings check OK                    
INFO[0000] Firewall status: ENABLED on "/sts-lab/host/sts-lab-cluster/192.168.10.20" 
INFO[0000] Firewall configuration OK on hosts:          
INFO[0000] 	"/sts-lab/host/sts-lab-cluster/192.168.10.20" 
INFO[0000] vCenter settings check OK                    
INFO[0000] License check OK on hosts:                   
INFO[0000]   "/sts-lab/host/sts-lab-cluster/192.168.10.20" 
INFO[0000] DRS check OK on:                             
INFO[0000]   "/sts-lab/host/sts-lab-cluster"            
INFO[0000]                                              
INFO[0000] Creating Resource Pool "vic-m1"              
INFO[0000] Creating appliance on target                 
INFO[0000] Network role "client" is sharing NIC with "public" 
INFO[0000] Network role "management" is sharing NIC with "public" 
INFO[0001] Creating directory [vic-datastore] VIC       
INFO[0001] Datastore path is [vic-datastore] VIC        
INFO[0002] Uploading images for container               
INFO[0002] 	"appliance.iso"                             
INFO[0002] 	"bootstrap.iso"                             
INFO[0012] Waiting for IP information                   
INFO[0025] Waiting for major appliance components to launch 
INFO[0026] Obtained IP address for client interface: "10.0.0.14" 
INFO[0026] Checking VCH connectivity with vSphere target 
INFO[0026] vSphere API Test: https://192.168.10.22 vSphere API target responds as expected 
INFO[0032] Initialization of appliance successful       
INFO[0032]                                              
INFO[0032] VCH Admin Portal:                            
INFO[0032] https://10.0.0.14:2378                       
INFO[0032]                                              
INFO[0032] Published ports can be reached at:           
INFO[0032] 10.0.0.14                                    
INFO[0032]                                              
INFO[0032] Docker environment variables:                
INFO[0032] DOCKER_HOST=10.0.0.14:2376                   
INFO[0032]                                              
INFO[0032] Environment saved in vic-m1/vic-m1.env       
INFO[0032]                                              
INFO[0032] Connect to docker:                           
INFO[0032] docker -H 10.0.0.14:2376 --tls info          
INFO[0032] Installer completed successfully   

can you help me how to generate the log bundle ?

**Comment by mdubya66** [comment]
also, you can get real time help from the community at https://vmwarecode.slack.com/ in the #vic-engine channel. You'll need to join first at https://code.vmware.com/join

And if you have a support contract you can contact vmware support through your normal escalation path.

**Comment by ibrahimabedalghafer** [comment]
Hi,

please find the attached log 
[port-layer.log](https://github.com/vmware/vic/files/1745622/port-layer.log)
[container-logs.zip](https://github.com/vmware/vic/files/1745623/container-logs.zip)
[logs.zip](https://github.com/vmware/vic/files/1745624/logs.zip)

**Comment by matthewavery** [comment]
@ibrahimabedalghafer Hi I am looking into this, I will try and update you as soon as possible!

**Comment by hickeng** [comment]
@ibrahimabedalghafer 

**TL;DR**
When using `run` without `-d` or `start -a` you are indicating that you wish to be attached to the container console and receive _all_ of the output from the process, without any truncation.
If using `start` without `-a` or `run -d` you are saying the container should launch in the background and you will attach to it later, receiving only the output from the point of attach onwards (`docker logs` will still return the full output for the container).

To just get a container running despite the current configuration issue you can use the following as these do not depend on an interactive connection:
```
docker run -dit -name test-container ubuntu /bin/my/command
docker logs test-container
```

**Details**
"Live" support is available at https://vmwarecode.slack.com/messages/vic-product

The way VIC accomplishes the former is by blocking the launch of the container process until the stdio connections have been established. This connection is initiated from the ESX server running the containerVM, with ESX acting as the client connecting to port 2377 on the VCH endpointVM over the management network.

There is a network diagram connections with ports and direction [here](https://vmware.github.io/vic-product/assets/files/html/1.3/vic_vsphere_admin/vch_networking.html#ports-and-protocols)

From the port-layer log I do not see any indication that the ESX host has successfully opened a TCP connection to the VCH endpointVM. This generally means that either:
1. the firewall on the ESX host is not permitting outbound to the endpointVM on 2377/tcp
2. the endpointVM is dropping the connection due to `rp_filter` because the ESX is not on the same management subnet

For (1), you should be able to log into the vicadmin page and see whether the `firewall` check is successful. [Details here](https://vmware.github.io/vic-product/assets/files/html/1.3/vic_vsphere_admin/vicadmin_status_ref.html#firewall).
Some additional sanity checking can be done by logging into the _ESX host_ (not vCenter) via `ssh` and using `nc` to confirm the connection, e.g:
```
[root@localhost:~] nc 192.168.78.127 2377
▒▒▒▒▒▒▒
[root@localhost:~]
```
The output will be unprintable as seen as it's a vSPC protocol negotiation, but the simple fact of connection is sufficient.

For (2) you need to specify additional routing destinations when configuring the management network gateway. [Details here](https://vmware.github.io/vic-product/assets/files/html/1.3/vic_vsphere_admin/mgmt_network.html#routing-destination-).

For extra detail, the following is a fragment of debug level portlayer logs for an incoming connection - I would expected to see the `INFO` level lines in your configuration but do not:
```
Feb 17 2018 17:47:50.047Z INFO  connection received
Feb 17 2018 17:47:50.047Z INFO  sending DO 232
Feb 17 2018 17:47:50.047Z DEBUG entered write loop
Feb 17 2018 17:47:50.047Z DEBUG Sending command: 253 232
Feb 17 2018 17:47:50.047Z INFO  sending DO 0
Feb 17 2018 17:47:50.047Z DEBUG Sending command: 253 0
Feb 17 2018 17:47:50.047Z INFO  sending DO 3
Feb 17 2018 17:47:50.047Z DEBUG Sending command: 253 3
Feb 17 2018 17:47:50.047Z DEBUG Sending command: 254 37
Feb 17 2018 17:47:50.066Z DEBUG Sending command: 253 232
Feb 17 2018 17:47:50.080Z DEBUG Sending command: 253 0
Feb 17 2018 17:47:50.080Z DEBUG Sending command: 253 3
Feb 17 2018 17:47:50.080Z DEBUG Sending command: 252 0
Feb 17 2018 17:47:50.080Z DEBUG Sending command: 252 3
Feb 17 2018 17:47:50.080Z INFO  vspc received KNOWN-SUBOPTIONS command
Feb 17 2018 17:47:50.080Z DEBUG [BEGIN]  [vic/lib/vspc.(*handler).knownSuboptions:130] handling KNOWN-SUBOPTIONS
Feb 17 2018 17:47:50.080Z DEBUG response to KNOWN-SUBOPTIONS: [255 250 232 1 0 1 2 3 40 41 43 44 45 46 48 70 71 73 80 81 84 85 86 87 82 83 255 240]
Feb 17 2018 17:47:50.080Z DEBUG [ END ]  [vic/lib/vspc.(*handler).knownSuboptions:130] [11.176µs] handling KNOWN-SUBOPTIONS
Feb 17 2018 17:47:50.080Z INFO  vspc received DO-PROXY command
Feb 17 2018 17:47:50.080Z DEBUG [BEGIN]  [vic/lib/vspc.(*handler).doProxy:147] handling DO-PROXY
Feb 17 2018 17:47:50.080Z DEBUG response to DO-PROXY: [255 250 232 71 255 240]
Feb 17 2018 17:47:50.080Z DEBUG [ END ]  [vic/lib/vspc.(*handler).doProxy:147] [30.786µs] handling DO-PROXY
Feb 17 2018 17:47:50.086Z INFO  vspc received VMUUID command
Feb 17 2018 17:47:50.086Z DEBUG [BEGIN]  [vic/lib/vspc.(*handler).cVMUUID:100] handling VMUUID
Feb 17 2018 17:47:50.086Z INFO  vmuuid of the connected containerVM: 52cefc17d082e3ee-8239b67475bce0bc
Feb 17 2018 17:47:50.086Z INFO  attempting to connect to the attach server
Feb 17 2018 17:47:50.086Z DEBUG [ END ]  [vic/lib/vspc.(*handler).cVMUUID:100] [104.812µs] handling VMUUID
Feb 17 2018 17:47:50.086Z DEBUG attach connector: Received incoming connection
Feb 17 2018 17:47:50.086Z DEBUG Initiating ssh handshake with new connection attempt
Feb 17 2018 17:47:50.086Z DEBUG HandshakeClient: Sending syn with pos 161
Feb 17 2018 17:47:50.086Z DEBUG Sending command: 254 44
Feb 17 2018 17:47:50.209Z DEBUG HandshakeClient: Sending ack with 109
Feb 17 2018 17:47:50.229Z DEBUG HandshakeClient: Connection established.
Feb 17 2018 17:47:50.229Z DEBUG HandshakeClient: connection handshake established
```
and then finally when the negotiation has succeeded we would see the notice of which container the connection is for and (if at debug level) some entries reporting that the container process has been unblocked:
```
Feb 17 2018 17:47:50.797Z INFO  Established connection with container VM: 6097fbb306e0fd39c6d1728c832e7b2f4c82bb483385d81fc27e5d3454b21964
Feb 17 2018 17:47:50.800Z DEBUG [BEGIN]  [vic/lib/portlayer/attach/communication.(*Connector).ids.func1:396] 6097fbb306e0fd39c6d1728c832e7b2f4c82bb483385d81fc27e5d3454b21964
Feb 17 2018 17:47:50.800Z DEBUG [BEGIN]  [vic/lib/portlayer/attach/communication.NewSSHInteraction:111] 6097fbb306e0fd39c6d1728c832e7b2f4c82bb483385d81fc27e5d3454b21964
Feb 17 2018 17:47:50.841Z DEBUG [ END ]  [vic/lib/portlayer/attach/communication.NewSSHInteraction:111] [41.506747ms] 6097fbb306e0fd39c6d1728c832e7b2f4c82bb483385d81fc27e5d3454b21964
Feb 17 2018 17:47:50.841Z DEBUG [ END ]  [vic/lib/portlayer/attach/communication.(*Connector).ids.func1:396] [41.541153ms] 6097fbb306e0fd39c6d1728c832e7b2f4c82bb483385d81fc27e5d3454b21964
Feb 17 2018 17:47:50.841Z DEBUG attach connector: Unblocking for 6097fbb306e0fd39c6d1728c832e7b2f4c82bb483385d81fc27e5d3454b21964
Feb 17 2018 17:47:50.841Z DEBUG [BEGIN]  [vic/lib/portlayer/attach/communication.(*interaction).Unblock:227]
Feb 17 2018 17:47:50.885Z DEBUG [ END ]  [vic/lib/portlayer/attach/communication.(*interaction).Unblock:227] [43.892266ms] 
Feb 17 2018 17:47:50.885Z DEBUG attach connector: Unblocked 6097fbb306e0fd39c6d1728c832e7b2f4c82bb483385d81fc27e5d3454b21964, returning
Feb 17 2018 17:47:50.885Z DEBUG attach connector: Found connection for 6097fbb306e0fd39c6d1728c832e7b2f4c82bb483385d81fc27e5d3454b21964: 0xc420a1e0f0
```
followed by some entries specifically for the stdio connection logic:
```
```

**Comment by ibrahimabedalghafer** [comment]
Dear hickeng,

I really appreciate your comment, but I checked the firewall rules like below :
the issue is solved by adding a static route in my wsxi host to the ip of vih machine.
I would to thank you all for your kind support, really appreciated.

BR
Ibrahim Abed-alghafer

**Comment by mdubya66** [comment]
Thank you @ibrahimabedalghafer. I'll close this issue. Please let us know if you have any other issues!
